### ISSUE TITLE
BufWriter incorrectly chunking UDP packets containing metrics

### ISSUE THREAD

**Issue by Geal** [sozu-proxy/sozu#481]
With the Clever Cloud deployment, we regularly get into a case where the metrics buffer is not large enough, so the `BufWriter` wrapping the UDP socket flushes the data in place, and we end up with some metrics cut in the middle, like this:

```
09:43:36.438890 IP n3-adc-cleverapps-clevercloud-customers.services.clever-cloud.com.45961 > n3-adc-cleverapps-clevercloud-customers.services.clever-cloud.com.8125: UDP, length 1496                                                                                    [53/1787]
E...<2@.@...................sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_02a2dcec-1b6b-4943-bfc9-188c6e01ef2a,backend_id=be5a5e94-4b81-4760-9cb9-72b757a8ac74:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_02a2dcec-1b6b-4943-bfc9-188c6e01ef2a,backend_id=be5a5e94-4b81-4760-9cb9-72b757a8ac74:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_03c7160e-16b8-47e7-97ff-e94b7bf76431,backend_id=4c58f59a-4dbc-4470-b707-5618bff8b271:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_03c7160e-16b8-47e7-97ff-e94b7bf76431,backend_id=4c58f59a-4dbc-4470-b707-5618bff8b271:0|c                                  
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_0537635f-407b-4d9e-8c76-75984e2bf6b3,backend_id=12011038-8696-43cb-9f8f-8b031430ec24:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_0537635f-407b-4d9e-8c76-75984e2bf6b3,backend_id=12011038-8696-43cb-9f8f-8b031430ec24:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_068d9e75-d809-446c-944d-7dbf7c4de667,backend_id=e62523a0-e541-44e7-abbe-2bd0353d05bd:0|c                                   
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_068d9e75-d809-446c-944d-7dbf7c4de667,backend_id=e62523a0-e541-44e7-abbe-2bd0353d05bd:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_0784d7a5-5892-4229-a483-efeb6f431943,backend_id=abf2607d-56f6-4357-a14c-7cbd154111b1:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_0784d7a5-5892-4229-a483-efeb6f431943,backend_id=abf2607d-56f6-4357-a14c-7cbd154111b1:0|c                                  
sozu.app.bytes_in,origin=WRK-04                                                                                                    
09:43:36.438934 IP n3-adc-cleverapps-clevercloud-customers.services.clever-cloud.com.45961 > n3-adc-cleverapps-clevercloud-customers.services.clever-cloud.com.8125: UDP, length 1491
E...<3@.@...................,version=0.7.2,app_id=app_0da8b321-2706-44a1-a95d-1f56b7741b56,backend_id=7f53c381-1fd4-470d-b680-05b81e95b094:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_0da8b321-2706-44a1-a95d-1f56b7741b56,backend_id=7f53c381-1fd4-470d-b680-05b81e95b094:0|c                                  
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_0da8b321-2706-44a1-a95d-1f56b7741b56,backend_id=e29eca39-420a-4840-8c8e-cb530dc6b7fb:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_0da8b321-2706-44a1-a95d-1f56b7741b56,backend_id=e29eca39-420a-4840-8c8e-cb530dc6b7fb:0|c          
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_0edc4b10-591a-4273-814b-112dcb841f5a,backend_id=6f67a034-f943-40b3-8821-cfd4f4f4a5aa:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_0edc4b10-591a-4273-814b-112dcb841f5a,backend_id=6f67a034-f943-40b3-8821-cfd4f4f4a5aa:0|c                                  
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_0fedcb5b-2fbd-4b4f-a4ee-e7177546bea4,backend_id=e86c949d-f9f8-4cce-bd59-c6f27649b285:179|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_0fedcb5b-2fbd-4b4f-a4ee-e7177546bea4,backend_id=e86c949d-f9f8-4cce-bd59-c6f27649b285:810|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_11cfa6fb-34fa-49b0-b7ef-1cbe55859141,backend_id=f03a9854-da07-4094-8f7e-8c11b4d69870:0|c                                   
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_11cfa6fb-34fa-49b0-b7ef-1cbe55859141,backend_id=f03a9854-da07-4094-8f7e-8c11b4d69870:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=                                                                                                      
09:43:36.438954 IP n3-adc-cleverapps-clevercloud-customers.services.clever-cloud.com.45961 > n3-adc-cleverapps-clevercloud-customers.services.clever-cloud.com.8125: UDP, length 1465
E...<4@.@...................app_11f605ea-507b-46c8-b7bf-8efc32babdb0,backend_id=34de61f1-3719-4298-926a-5763731d1d9f:0|c                                                            
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_11f605ea-507b-46c8-b7bf-8efc32babdb0,backend_id=34de61f1-3719-4298-926a-5763731d1d9f:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_12468da9-866c-4a29-aa9c-82965743f4c4,backend_id=c991b1b7-328d-4cfd-9a28-eb547f0665fa:0|c          
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_12468da9-866c-4a29-aa9c-82965743f4c4,backend_id=c991b1b7-328d-4cfd-9a28-eb547f0665fa:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_1246f211-d4a7-4787-ba62-56c163a8b4ef,backend_id=428c5613-cc2b-4a33-b74b-ceb4c185daab:0|c                                   
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_1246f211-d4a7-4787-ba62-56c163a8b4ef,backend_id=428c5613-cc2b-4a33-b74b-ceb4c185daab:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_1246f211-d4a7-4787-ba62-56c163a8b4ef,backend_id=da616112-47ad-402f-8d3f-57a74d98e596:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_1246f211-d4a7-4787-ba62-56c163a8b4ef,backend_id=da616112-47ad-402f-8d3f-57a74d98e596:0|c                                  
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=app_13bf0a68-c341-498b-8c80-c158e67e0937,backend_id=990fb109-7ce1-4f39-8622-f319dae135dc:0|c
sozu.app.bytes_out,origin=WRK-04,version=0.7.2,app_id=app_13bf0a68-c341-498b-8c80-c158e67e0937,backend_id=990fb109-7ce1-4f39-8622-f319dae135dc:0|c
sozu.app.bytes_in,origin=WRK-04,version=0.7.2,app_id=   
```
6182d3107a7d14cf33459 contains a work around (we flush the buffer once we get 10 metrics in there), but it's not a future proof solution.

Here's what I would like to see:
- something like a `BufLineWriter` implementation, that would still support `write_fmt`, but if there's not enough room in the buffer, it would flush all of the buffered data except what is written with the current call to `write_fmt`
- some configuration option for the buffer's capacity, because it depends on the maximum UDP packet size accepted by the kernel, the network and the metrics server
- some work to avoid sending `count` metrics if they were not modified recently

**Comment by Geal** [comment]
hello, there's a temporary fix in 5349227. It is temporary because it would result in one packet sent for every metric, which could be too much traffic. The `BufLineWriter` idea is still interesting as a proper fix.
But with this temporary fix we'll at least get all the metrics.

**Comment by Geal** [comment]
I rewrote again the linewriter, haven't seen much issues yet
